### ISSUE TITLE
FPS drop DragonAPI 13b-16d

### ISSUE THREAD

**Issue by SvEgiiVEteR** [ReikaKalseki/Reika_Mods_Issues#857]
http://i.imgur.com/T7F5nh4.png
http://i.imgur.com/OxXCOup.png
drops fps when looking down... happens when i install DragonAPI 1.7.10 V13b.jar
here is the config
https://gist.github.com/7e6a017ef2c59ae5dd5a9d0ddd4d96e4
mod
https://gist.github.com/f888ece36ecee702d8d49eb74b932088

without DragonAPI
http://i.imgur.com/wvTy1kB.png
http://i.imgur.com/aWxuk27.png

and only mod DragonAPI 13b:
look down 30-40 fps
look up 150 fps

and only mod DragonAPI 12e:
look down 160-180 fps
look up 230 fps

**Comment by lee321987** [comment]
I have the same problem.  About %50 FPS drop going from 12e to 13b (only mod).
Do you have a Radeon HD 4000 series graphics card?
I have HD 4670.

**Comment by SvEgiiVEteR** [comment]
I have HD 5770.
leaves a lot of fps when looking down

**Comment by ReikaKalseki** [comment]
Use VisualVM to get profiling data.

**Comment by SvEgiiVEteR** [comment]
this tab is?
http://i.imgur.com/MIEWfiz.png
https://mega.co.nz/#!6Y8EXDCA!fw3N-Dyib6YmARh6DzW5YQSmcPmxbAhgtx5oT0mcte0

**Comment by lee321987** [comment]
I used the Sampler tab, and Reika didn't tell me that was wrong.
And make sure you're using DragonAPI v13 when you do it.

**Comment by SvEgiiVEteR** [comment]
and Sampler
https://mega.co.nz/#!3J9UXCSI!iX7G-zJqqpGyRFRu1eDS6-_mduPNHshzO2PdM-gC_Ts

**Comment by ReikaKalseki** [comment]
I need a site that actually works.

**Comment by SvEgiiVEteR** [comment]
https://dl.dropboxusercontent.com/u/26268735/ShareX/2016/05/prof.zip
https://dl.dropboxusercontent.com/u/26268735/ShareX/2016/05/profS.zip

**Comment by ReikaKalseki** [comment]
Run a longer sample, and what utility do you think I would get from this?

![](http://i.imgur.com/6i31wIy.png)

**Comment by SvEgiiVEteR** [comment]
the light was turned off at home now.. will do .. the CPU?
I don't know why 300 ms. 15 minutes I created the snapshot
https://dl.dropboxusercontent.com/u/26268735/ShareX/2016/05/simpler.rar
1 888 000 ms + threaddump
or https://dl.dropboxusercontent.com/u/26268735/ShareX/2016/05/snapshot-cpu-calltree.rar
5 999 000 ms
look down 30-41 fps
and log https://gist.github.com/7563f72283fac73b69047f20b6072556

**Comment by SvEgiiVEteR** [comment]
tried DragonAPI 1.7.10 V14a.jar FPS drops too

**Comment by lee321987** [comment]
Same here -- V14a same FPS drop.

**Comment by SvEgiiVEteR** [comment]
tried DragonAPI 1.7.10 V14с.jar FPS drops too :(

**Comment by ghost** [comment]
Confirmed this. Only use Thermos + DragonAPI.

**Comment by ReikaKalseki** [comment]
> Only use Thermos + DragonAPI.

**Comment by SvEgiiVEteR** [comment]
my tests FPS were in the single player game.

**Comment by ghost** [comment]
@ReikaKalseki @SvEgiiVEteR Oh...I test it in Server(Just a simple test, similar to the stand-alone. This is no different with the single...). But I think we can use some optimization mod like fastcraft, betterfps you know. **This only occurs when I only installed DragonAPI.**

**Comment by SvEgiiVEteR** [comment]
@AuthMe fastcraft, betterfps - they did not help me. fps drops

**Comment by Aivech** [comment]
Does it happen in a superflat with the default superflat preset?

**Comment by SvEgiiVEteR** [comment]
@DaMachinator no. the default generation of the world.

**Comment by Aivech** [comment]
So it doesn't happen on a superflat?

**Comment by SvEgiiVEteR** [comment]
@DaMachinator DragonAPI 1.7.10 V14d.jar. look down and fps drops to 36(default generation)
superflat 180 fps...

DragonAPI 1.7.10 V15a
look down 30-41 fps
((((((((((

**Comment by lee321987** [comment]
This may actually be caused by an IMPROVEMENT (less CPU requirements) in DragonAPI, so your CPU is throttling down.
Try forcing your CPU to stay in it's lowest P-state (full speed), by disabling Cool & Quiet in your BIOS, or if you use MsrTweaker click the "Service" button and set the "P-state bounds" to 0 and 0.

**Comment by Aivech** [comment]
It may also help to turn on "Advanced OpenGL" in your MC settings if you have that option.

**Comment by SvEgiiVEteR** [comment]
@DaMachinator if "Advanced OpenGL" on. I have sometimes crashes the client...

**Comment by Aivech** [comment]
Decidedly odd, that.
Nonetheless I do have one more theory, that should be testable by running a lightly modded install on very bad hardware, as that should exacerbate it so that it is noticeable without DragonAPI if my theory is correct.

My theory is:
You ALWAYS get an FPS drop when looking down. There's more stuff for the computer to render down there - caves, etc. However, normally the drop is too small to notice - until you install a large mod like DragonAPI.
If this is true, the drop should also be noticeable on a vanilla, vanilla + forge only, or lightly modded install run on much worse hardware.

**Comment by ReikaKalseki** [comment]
Advanced OpenGL was specifically designed to avoid that.

**Comment by rwizard73** [comment]
@DaMachinator I will be able to test your hypothesis to some extent in the next couple of days. However, this issue is not noticeable using DragonAPI v12, only using v13 or later so I doubt that this is the issue in any case. Unless v12 is so much less complicated that this goes from not noticeable at all to severely noticeable?

**Comment by Aivech** [comment]
@ReikaKalseki I know it was designed to avoid rendering objects you can't actually see.
However:
1. It doesn't always work properly.
2. That option is not always available.

**Comment by lee321987** [comment]
Maybe we should all post certain info about our computers, so we can find out what we (who have this problem) have in common.
Here's the things I think we should list:
Operating system
CPU
GPU
Anti-Virus

Can anyone think of anything else?

**Comment by lee321987** [comment]
How about HPET?  Anyone else have their HPET on in the BIOS?

**Comment by SvEgiiVEteR** [comment]
16a too

**Comment by SvEgiiVEteR** [comment]
very bad fps (((

**Comment by Aivech** [comment]
1. 16c is the latest version of Reika's Mods at present.
2. Make sure you have the **latest version** of FastCraft. If you do, try removing it.
3. If your graphics card supports this, try enabling Advanced OpenGL. This may or may not help. As far as I'm concerned, it runs on magic smoke.

Note that looking down will ALWAYS cause an FPS drop, unless you have your FPS capped and it's still above the FPS cap. There are more blocks to render on and under the ground.

**Comment by SvEgiiVEteR** [comment]
@DaMachinator 
1 DragonAPI 1.7.10 V16d
2 also installed and removed(mod improves fps, but DragonAPI all breaks )
3 included did not help,so I can play without OptiFine(makes worse fps), but without him the options that I have.

I have an Assembly DragonAPI 12 130 fps(looking down), mods 200
everything is OK.
on the Assembly with 3 modes
DragonAPI 1.7.10 V16d.jar
InGameInfoXML-1.7.10-2.8.1.82-universal.jar
LunatriusCore-1.7.10-1.1.2.21-universal.jar

20-40 fps(looking down) if up in the air 20 fps
the author of the mod that broke with the above 12 version, and it seems unable or unwilling to fix...

UP
fastcraft-1.23.jar + OptiFine_1.7.10_HD_U_D4.jar + enabling Advanced OpenGL
it was 20-40 fps(looking down) if up in the air 20 fps
now 50-60 fps(looking down) if up in the air 58 fps

but as you can see it's just a crutch. with 3 modes available.

DragonAPI 12 
190 fps(looking down) if up in the air 190 fps
DragonAPI 12 + enabling Advanced OpenGL
180 fps(looking down) if up in the air 180 fps

**Comment by Builderb0y** [comment]
> Note that looking down will ALWAYS cause an FPS drop, unless you have your FPS capped and it's still above the FPS cap. There are more blocks to render on and under the ground.

False. Minecraft doesn't render any chunks outside the view frustum, and solid blocks that are adjacent to other solid blocks get their faces culled. As such, you'd be rendering a small amount of grass, a couple caves, and maybe some lava if you're lucky. More than 99% of the blocks you're looking at simply won't be rendering at all.

**Comment by gartia** [comment]
I seem to be getting a fps halving when i add dragon api. Is this a known bug or could it be other mods i have conflicting. I thought it was a different mod of reika's, but i tested it just by adding dragonapi alone and that's what the cause was.

**Comment by DarianLStephens** [comment]
I never noticed this in earlier versions (Possibly because I wasn't paying attention), but for some reason, all blocks that are normally things like 'minecraft:blocks.dirt' are now 'dragonapi:minecraft:blocks.dirt'. I'm sure there's some important technical reason for that, and it may not be related at all to the fps, and if it is, it's unlikely to change.

**Comment by Builderb0y** [comment]
1: That only applies to technical blocks, like off-state torches/repeaters.
2: That's done so that reika can render said blocks in his GUIs, because he doesn't want to make a separate texture for all of them. For example, the magnetostatic GUI has a button in it that cycles between high, low, and ignored redstone control. These states are represented by redstone torches, off-state redstone torches, and redstone dust respectively.
3: That's NOT the cause of this issue, but you are correct that it's unlikely to change.

**Comment by DarianLStephens** [comment]
Well, I'd be happy to provide any information I can to help resolve this issue, since it's crippling usability when FPS fluctuates so wildly; it makes me queezy.

**Comment by SvEgiiVEteR** [comment]
my fps drops in half since the dragonapi. on a clean map, with 1 this mod.
12e ok

**Comment by lee321987** [comment]
Well I got a new computer, and this is still an issue.
I flew up to Y200 (view distance 10), and looked straight down.
Tested using only mod DragonAPI.
As the chunks are first loading, FPS stays around what it is with no mods (75 FPS), till about 75% (guessing) of the chunks have loaded, then it starts gradually going down, settling at 42 FPS before all chunks have loaded.

CPU: i5-6600
GPU: Radeon RX 480

**Comment by lee321987** [comment]
I may have found a solution.
I've never use ChromatiCraft (CC), but I installed it just to see if it helped.
On first run it seemed to stall the game for a few minutes (found out DragonAPI was "downloading remote assets").
After starting a world that included CC, I uninstalled CC, and now new worlds seem to have normal FPS.

List of mods used during this test:

DragonAPI 1.7.10 V16d.jar
ChromatiCraft 1.7.10 V16d.jar
RotaryCraft 1.7.10 V16d.jar
ReactorCraft 1.7.10 V16d.jar

Batty's Coordinates PLUS Mod for Forge-1.7.10_1.6.0.jar
fastcraft-1.23.jar
OptiFine_1.7.10_HD_D1.jar

**Comment by ReikaKalseki** [comment]
Wait, so adding and removing CC fixes the issue....? Does this persist across reloads?

**Comment by lee321987** [comment]
Sorry Reika, I had Vsync on, and it messed up my tests.  (so i just deleted the whole comment)

**Comment by Joe12o** [comment]
Running into this issue as well, went from 400-500 fps without to 60-120 fps with.

Here's some profiler results both with:
![image](https://cloud.githubusercontent.com/assets/1216155/24391146/646f9698-135b-11e7-8a92-8a5dbd92dc12.png)

and without DragonAPI:
![image](https://cloud.githubusercontent.com/assets/1216155/24391250/ed978584-135b-11e7-8efc-de335703106a.png)

Only other mods used were optifine and fastcraft with the only difference between the two results being whether or not DragonAPI was installed. Note that `func_78419_a()` is actually `net.minecraft.client.renderer.RenderList.callLists()`. It seems to take a marginal amount of time without DragonAPI and more time than everything else combined with DragonAPI.

**Comment by Joe12o** [comment]
Some more results with just DragonAPI:

Run 1:
![image](https://cloud.githubusercontent.com/assets/1216155/24629338/af49e206-1886-11e7-8ca5-7d550884d187.png)

Run 2:
![image](https://cloud.githubusercontent.com/assets/1216155/24629448/03d370c6-1887-11e7-997d-220c59ea9ac6.png)

Different results than last time for some reason and no idea what could be causing it, just that it _is_ DragonAPI. Tried looking into it myself, but it seems like DragonAPI is _way_ more invasive than I expected an "API" to be so I'm not too willing to look over all the changes. Plus, if this for sure happened in the 12e-13b transition, it should be relatively easy for Reika to go over the changes and see which caused it. I would totally try that myself too, but the infrequent, large, and cryptically named commits make it a pain. If this happens to every user it seems like this should be way higher priority than it seems like it is currently... 😞

... @ReikaKalseki

**Comment by OvermindDL1** [comment]
Wow I forgot just how inefficient normal MC rendering is, they really do not know how to make a rendering engine...

Either way, your stack traces imply two things:

1. First that an anomalous call is being injected via ASM with the name of `doRender`, which is really slow, it seems it is being done from:  https://github.com/ReikaKalseki/DragonAPI/blob/19a41c8f323c73b43a25dc3344ddd49dc8d391b3/ASM/Patchers/Hooks/Event/Render/EntityRender.java

2. Second, that the doRender call is slow because (assuming you have *nothing* but forge in the first image and *nothing* but forge and DragonAPI in the second) the OpenGL state is being left in an error state considering the `GL11.glGetError()` time taken.  This would be hard to trace down because it might not actually be 'in' doRender but would be caused by something rendering, however if all you have is just forge and dapi then it is likely a dapi opengl call (of which it has many) is not being cleaned up properly.  It could also be caused by an excessive amount of OpenGL state push/pops instead of altering state smartly, but this would have to be pretty excessive to be that big anyway.

Would it be possible to attach an AMD or NVidia OpenGL debugger (depending on the video card you have) and report back and state errors it reports, or just in general dump as much information as is useful from it?

And yes, this slow-down happens to me as well with Reika's mods (though I've not tried just DAPI by itself), but I'm unable to test at moment, so if you could please @Joe12o .  :-)

**Comment by Joe12o** [comment]
If you can point me to an NVIDIA OpenGL debugger, I'd be happy to. I've used gDEBugger in the past, but it seems like it died off. @OvermindDL1

**Comment by OvermindDL1** [comment]
@Joe12o As long as it supports your drivers then gDEBugger should be fine, MC's OpenGL is *ancient* anyway.  Otherwise the official NVidia one is:  https://developer.nvidia.com/nvidia-nsight-visual-studio-edition

Don't ask me how to use it though, I run AMD.  ^.^

**Comment by Joe12o** [comment]
@OvermindDL1 Ah, alright. Might get to this tomorrow if I find the time for it. Looks like CodeXL is the successor to gDEBugger so I downloaded that, but need to figure out how to make it attach to a LWJGL process.

**Comment by ReikaKalseki** [comment]
> however if all you have is just forge and dapi then it is likely a dapi opengl call (of which it has many)

DragonAPI on its own has very little rendering.

**Comment by HellFirePvP** [comment]
@Joe12o Does your GPU run on multiple threads? (normally that's the driver ultimately controlling that _iirc_)

If so, error checks get more and more expensive as the execution has to wait until all threads are synced up. That's also why you _normally_ don't ship glGetError calls into release versions of software that uses OpenGL as it is a potentially very expensive check depending on your hardware/driver. Of course though, due to the instability of modded minecraft, i wouldn't leave it out.

If your GPU driver supports multithreading, check if it is enabled and/or if there's a way to specifically disable that - If there is a way to disable that, please try do so and rerun the profiling as you did above.

**Comment by OvermindDL1** [comment]
> DragonAPI on its own has very little rendering.

Very little rendering, but it controls a lot of rendering state I saw yep.

> If your GPU driver supports multithreading, check if it is enabled and/or if there's a way to specifically disable that - If there is a way to disable that, please try do so and rerun the profiling as you did above.

Perhaps, but it did not show up in the non-DAPI run?

**Comment by Joe12o** [comment]
Seems like [this](https://github.com/ReikaKalseki/DragonAPI/blob/master/Auxiliary/DragonAPIEventWatcher.java#L185) call to glPushAttrib() is doing it, even if disableAlphaTest is set to false in the config. glPushAttrib() is generally pretty slow (especially with GL_ALL_ATTRIB_BITS) so it'd probably be good to avoid it where possible, but I wouldn't expect it to cause this kind of slowdown so I'm not sure that its the root cause of the issue. Completely removing both disableAlphaTest() methods fixes the symptoms of the issue on my end, but you'll likely want to find a better way (such as resetting the glAlphaFunc manually instead of needlessly saving/restoring the entire state) as I'm not too sure of the implications of removing it with your mods @ReikaKalseki.

You'll also need to fix it [here](https://github.com/ReikaKalseki/ChromatiCraft/blob/master/ChromaClientEventController.java#L309) in chromaticraft as it causes the same issue to occur. Again, completely removing both methods solves it on my end, but it'll likely cause some unexpected behavior in the future.

**Comment by HellFirePvP** [comment]
Hmm, i'm uncertain that this would cause that much of an overhead. Even with pushing all attribute states, it doesn't normally take that much time.. Though, doing this just for the sake of doing it seems like unnecessary overhead.

**Comment by Joe12o** [comment]
Yeah, it's possible though because iirc RenderWorldEvent is fired for each 16x16x16 chunk being drawn so the overhead adds up at a high enough view distance. Also, I did notice that my fps dropped as more and more chunks loaded in.

**Comment by HellFirePvP** [comment]
That is indeed true though. It *is* fired for each 16x16x16 area being drawn.

**Comment by ReikaKalseki** [comment]
@Joe12o : You have at least found some code that was added around the time of v13's release, thus making it a far better candidate than anything else thus suggested.

That push is not
> for the sake of doing it

it is because several CC block renders (simple textures, not even ISBRHs) glitch out if I do not disable GL_ALPHA_TEST before the WorldRenderer runs. This was also later used for Void Gaze.

Also, I highly doubt pushing or popping the matrices is unperformant, especially given the fact that they are strongly recommended over manually resetting each flag (lighting, face culling, blending, etc) manually.

**Comment by OvermindDL1** [comment]
> Hmm, i'm uncertain that this would cause that much of an overhead. Even with pushing all attribute states, it doesn't normally take that much time.. Though, doing this just for the sake of doing it seems like unnecessary overhead.

It easily could, MC's Entity rendering is *amazingly* inefficient, and that happened to be getting called during those passes, so I could see how it could.  And pushing all bits is pretty inefficient on many drivers regardless, it would be better to cache it in gl objects, but you need OpenGL 2.1+ for that, and MC is, well, crap...

**Comment by HellFirePvP** [comment]
So, well, TIL that pushing all attribs takes at *best* performance around 30-40 microseconds; Not surprising due to the amount of states that are getting saved and later on loaded.

Also, yes, modern graphics cards and their drivers are no longer optimized on that as the attribute stack and everything regarding that got removed from OpenGL 3.X onwards and 2.1 is already ancient...

Edit: https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glPushAttrib.xml
A detailed list of states being saved on the attribute stack.

**Comment by Joe12o** [comment]
> Also, I highly doubt pushing or popping the matrices is unperformant, especially given the fact that they are strongly recommended over manually resetting each flag (lighting, face culling, blending, etc) manually.

From what I've seen, pushing/popping the entire attribute stack is pretty slow in all situations. Plus, depending on driver implementation, glPopAttrib() might assume that every attrib changed so it might cause the driver to flush the pipeline. Also, I'm not sure who strongly recommended them, but setting/resetting each attrib manually will always be faster because you only change the attribs you need to. As long as you don't forget to reset them, there is no difference other than performance.

**Comment by kutschkem** [comment]
GL_COLOR_BUFFER_BIT contains the alpha test bit and function according to https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glPushAttrib.xml

Maybe try just pushing / popping that instead of everything?

**Comment by SvEgiiVEteR** [comment]
I installed all Mods Reika v17 and nothing that is not installed.
On average 50 FPS in the air. 
it's better than it was at 16.
but again, it's worse than v12

v13
look down to 30-40 fps
look up 150 fps

v17:
look down to 60-80 fps
look up 160 fps

v12:
look down 160-180 fps
look up 230 fps

I don't know how it would work with a full pack of my mods...

**Comment by DarkShadow44** [comment]
I can confirm the issue, and that stubbing disableAlphaTest like Joe12o  suggested massively improves the FPS.
Using DragonAPI 1.7.10 V17c.

**Comment by SvEgiiVEteR** [comment]
@DarkShadow44 
Disable Alpha Clipping in WorldRenderer
you mean this option? to turn it on or off?
Yes 17 version was better than 13-16
but the 12 version was the most stable for FPS

Disable Alpha Clipping in WorldRenderer"=true
look down to 110-125 fps
Disable Alpha Clipping in WorldRenderer"=false
look down to 100-120 fps
for me, the difference is barely noticeable

**Comment by DarkShadow44** [comment]
No, I mean editing Reika/DragonAPI/Auxiliary/DragonAPIEventWatcher.java to make both disableAlphaTest functions do nothing.

Without DragonAPI: ~120FPS
With DragonAPI: ~10FPS
With patched DragonAPI: ~125FPS

With Optifine:
Without DragonAPI: ~160FPS
With DragonAPI: ~12FPS
With patched DragonAPI: ~160FPS

**Comment by SvEgiiVEteR** [comment]
@DarkShadow44 and where to get patched.. I would like to check

**Comment by DarkShadow44** [comment]
@SvEgiiVEteR 
You need to download and compile it from https://github.com/ReikaKalseki/DragonAPI.
Without permission from Reika, I can't distribute a patched version.

**Comment by DarianLStephens** [comment]
What's the performance like if, instead of blanking the functions, we change them to just push/pop the attributes Reika wants instead of the whole stack?

**Comment by DarkShadow44** [comment]
@DarianLStephens Which flags would I need to use for glPushAttrib then?

**Comment by DarianLStephens** [comment]
If Reika's problem is with transparency, probably something like the GL_ALPHA_TEST flag.
Reika did this due to some issue rendering one or multiple Chromaticraft blocks, but I'm not sure which.
If I did, then you could use it to see if it still renders correctly.

**Comment by ReikaKalseki** [comment]
> What's the performance like if, instead of blanking the functions, we change them to just push/pop the attributes Reika wants instead of the whole stack?

That, if it works and breaks nothing else, will be implemented natively.

**Comment by DarianLStephens** [comment]
I've been trying to set up a Minecraft dev environment for a few days now, and have had zero success. I know this is somewhat off-topic, but DragonAPI seems like one of the more difficult things to set up an environment for, and thus compile.
I'm interested in tinkering with the alphaTest functions, but without being able to compile it, I can't do anything.

**Comment by ReikaKalseki** [comment]
> DragonAPI seems like one of the more difficult things to set up an environment for, and thus compile.

This is intentional (to prevent others from releasing illegitimate versions); there is no way you will be able to do it.

What would the exact code changes be?

**Comment by DarianLStephens** [comment]
I've not fiddled with stuff like this before, but I would like to compile a version without the alphaTest things as a performance baseline, then maybe replace the 'GL11.glPushAttrib(GL11.GL_ALL_ATTRIB_BITS)' of 'DragonAPIEventWatcher.java' with 'GL11.glPushAttrib(GL11.GL_ALPHA_TEST)', and then adding the 'GL_ALPHA_TEST' argument to 'GL11.glPopAttrib();' in the next code block.
I don't know if that would work, or help anything, which is why I didn't want to bother anyone with this while I tested it and learned myself.
If you were to allow me to compile it, I have no interest in distributing it to anyone but yourself, should you ask for it. I'm just interested in solving this performance issue, because it does seriously hamper performance, from constant 60fps to <30. I'm not even using Chromaticraft, just LegacyCraft and ExtendedRedstone, and the latter is too ingrained in my world to remove now.

**Comment by ReikaKalseki** [comment]
Do you know ASM? That would be an easy way to modify the code at runtime.

**Comment by OvermindDL1** [comment]
> This is intentional (to prevent others from releasing illegitimate versions); there is no way you will be able to do it.

I still don't see how that would stop anyone.  Anyone could just drop it into a new forge workspace then setup the deps.  Heck, even if some code is missing then they could just decompile the mod jar itself into a fully compilable workspace dead-simple nowadays.  Only thing it does is just make people not want to help fix bugs.

**Comment by ReikaKalseki** [comment]
>  Anyone could just drop it into a new forge workspace then setup the deps. Heck, even if some code is missing then they could just decompile the mod jar itself into a fully compilable workspace dead-simple nowadays.

No, it is nowhere near that simple.

**Comment by OvermindDL1** [comment]
I've decompiled quite a variety of mods into perfectly compilable code (usually only minor hand-edits needed if any, modern decompilers are quite robust), in addition to just editing the bytecode out-right (really not that hard at all).

**Comment by ReikaKalseki** [comment]
That may well be true, but my mods are built in a *very* different manner.

**Comment by FuYang001** [comment]
Any idea to solve this problem?

**Comment by ReikaKalseki** [comment]
https://www.reddit.com/r/feedthebeast/comments/7rswqn/fastcraft_2_preview/dt088gd/

**Comment by TheCupcakeisalie** [comment]
Getting a 75% fps drop, I don't see how anything in that thread helps me solve my problem, as I can't figure out how to use ASM to remove the methods.

**Comment by DarkShadow44** [comment]
Well, in theory I could give you a patched version, but I'd need permission for that.
I could also give you detailed instructions how to apply my workaround, if you want that, but it's a bit advanced.

**Comment by TheCupcakeisalie** [comment]
Did you make a coremod to do bytecode manipulation? I have been trying to figure this out for hours now and would love some help

**Comment by ReikaKalseki** [comment]
v20 has not yet been released, so the potential fix is not yet available.

**Comment by DarkShadow44** [comment]
@TheCupcakeisalie
Okay, I forgot how hard this mod is to edit. I guess we'll have to wait for an official fix.
I mean, it's doable, but it's complex and I don't think I should elaborate on how to work around the mods security here, so I won't.

**Comment by lee321987** [comment]
@ReikaKalseki
I could have sworn you said you might make some option in the DragonAPI config for this issue.  Or maybe something about a JVM argument.  Did you change your mind?

**Comment by Mikki-m8** [comment]
@lee321987 
I don't think there's any config option or JVM argument, but I can say that my performance is much better with v20. Before I could barely run a pack with just Reika's mods and utility mods, but now I have them in 270 mod pack with no discernible performance loss, aside from worldgen. Might just be a coincidence, but I'm happy.

**Comment by ReikaKalseki** [comment]
@lee321987 That got implemented. It is a JVM argument.

**Comment by DarianLStephens** [comment]
Well... what's the argument? I haven't found any sign of it.

**Comment by ReikaKalseki** [comment]
`-DragonAPI_ForceAlphaClip`

**Comment by DarianLStephens** [comment]
That seems to have done it; I'm back to a near-constant 60fps.

**Comment by ReikaKalseki** [comment]
Well, then...

**Comment by p3nguinPinata** [comment]
Wanted to let you know that the argument worked for me as well. Previously I would have 125~ fps with shaders turned off, and around 22fps with shaders on. After plugging in the JVM argument from above my fps are hanging out around 60.

**Comment by TheCupcakeisalie** [comment]
On the latest version of DragonAPI the JVM arg doubles my performance with just DragonAPI installed on a fresh world.  Installing ChromatiCraft halves my performance, just as removing the JVM arg halves my performance.  With ChromatiCraft installed the and JVM arg enabled the performance difference is negligible.   I suspect ChromatiCraft is not this heavy and that this is a continuation of the glPushAttrib(GL11.GL_ALL_ATTRIB_BITS) issue, but I may be wrong.  Thoughts?

**Comment by ReikaKalseki** [comment]
CC adds a lot of render handlers, many of which are going to be pushing and popping attrib matrices.

**Comment by Skyler-Altol** [comment]
I'm going to weigh in on this and state that this JVM argument has made my frame rate go from a stable 67-69 all the way up to 300+ stable in an identically generated fresh world. Whatever that argument is doing, it's having a *significant* improvement to performance. If it's possible to apply this to CRC, **please** look into it :D

I have ROC, REC, ELC and DAPI installed, and this arg results in my performance being identical to when they're not installed. Unfortunately I can confirm @p3nguinPinata's findings regarding CRC bringing the fps back down to pre-argument levels. So for now, I will not be using CRC.
